core/src/main/java/org/apache/calcite/plan/Digest.java

```diff
@@ -0,0 +1,269 @@
+  }
+
+  /**
+   * Collects the items used for {@link #hashCode} and {@link #equals}.
+   *
+   * <p>Generally, the items used for hashCode and equals should be the same. The exception
+   * is the row type of the relational expression: the row type is needed because during
+   * planning, new equivalent rels may be produced with changed fields nullability
+   * (i.e. most of them comes from the rex simplify or constant reduction).
+   * This expects to be rare case, so the hashcode is computed without row type
+   * but when it conflicts, we compare with the row type involved(sans field names).
+   *
+   * @param rel      The rel to compute digest
+   * @param contents The rel properties should be considered in digest
+   * @param withType Whether to involve the row type
+   */
+  private static Object[] collect(
+      RelNode rel,
+      List<Pair<String, Object>> contents,
+      boolean withType) {
+    List<Object> hashCodeItems = new ArrayList<>();
+    // The type name.
+    hashCodeItems.add(rel.getRelTypeName());
+    // The traits.
+    hashCodeItems.addAll(rel.getTraitSet());
+    // The hints.
+    if (rel instanceof Hintable) {
+      hashCodeItems.addAll(((Hintable) rel).getHints());
+    }
+    if (withType) {
+      // The row type sans field names.
+      RelDataType relType = rel.getRowType();
+      if (relType.isStruct()) {
+        hashCodeItems.addAll(Pair.right(relType.getFieldList()));
+      } else {
+        // Make a decision here because
+        // some downstream projects have custom rel type which has no explicit fields.
+        hashCodeItems.add(relType);
+      }
+    }
+    // The rel node contents(e.g. the inputs or exprs).
+    hashCodeItems.addAll(contents);
+    return hashCodeItems.toArray();
+  }
+
+  /** Normalizes the rel node properties, currently, just to replace the
+   * {@link RelNode} with a simple string format digest. **/
+  private static List<Pair<String, Object>> normalizeContents(
+      List<Pair<String, Object>> items) {
+    List<Pair<String, Object>> normalized = new ArrayList<>();
+    for (Pair<String, Object> item : items) {
+      if (item.right instanceof RelNode) {
+        RelNode input = (RelNode) item.right;
+        normalized.add(Pair.of(item.left, simpleRelDigest(input)));
+      } else {
+        normalized.add(item);
+      }
+    }
+    return normalized;
+  }
+
+  /**
+   * Returns a simple string format digest.
+   *
+   * <p>There are three kinds of nodes we need to handle:
+   *
+   * <ul>
+   * <li>RelSubset: composition of class name, set id and traits;</li>
+   * <li>HepRelVertex: composition of class name and current rel id;</li>
+   * <li>Normal rel: composition of class name and id.</li>
+   * </ul>
+   *
+   * @param rel The rel
+   */
+  private static String simpleRelDigest(RelNode rel) {
+    StringBuilder digest = new StringBuilder(rel.getRelTypeName());
+    digest.append('#');
+    if (rel instanceof RelSubset) {
```

Comment:
Why not let the `rel` node itself handle how to create the digest vs having a non-extendable/overridable method using different behaviors based on the type of `rel` to do it? This looks like an anti design pattern.

Reply:
For current impl, the digest **is** handled by each `rel` node. Split the code everywhere is hard to maintain, let the `Digest` take care how the digest should be generated but the invoker is still each node.

Reply:
Considering that method, the digest takes care how the digest should be generated for nodes it already knows about. Based on the current approach, if I create a new RelNode subclass in my project which has a very natural/efficient way of creating a digest, I would have no choice than to fork the Calcite project to modify that method.
I would favor an approach which would let `HepRelVertex` itself decides what to use for its digest instead of hardcoding in this method that the current rel should be chosen.

Reply:
> RelNode subclass in my project which has a very natural/efficient way of creating a digest, I would have no choice than to fork the Calcite project to modify that method
Did you notice that the `#computeDigest` and `#explainTerms` are both `API` for overriding the digest ?

`HepRelVertex` is not a public API, i don't think there is necessity to customize its digest, if you have a sub-class, override the `#computeDigest` is till feasible.

Reply:
why not using the {{#computeDigest}} with the vertex id then?

Also it is not possible to "substract" to the default digest. One can only add. The patch you proposed has some behavior which is not possible to change

Reply:
The current patch also supports pure string digest, which is same as before, if you think it is impossible, please give an example.

The digest of `HepRelVertex` and `RelSubSet` are the same as before, what do you mean by `using the #computeDigest with the vertex id then` ?

Reply:
I guess I missed some update to your change (or missed the second constructor), but it looks like the choice is either:
- provide rel + a list of extra items, and the logic to create the hashcode + digest is totally controlled by the digest class
- provide rel + digest string to skip the whole hash computation logic

I guess nothing in-between...

> The digest of HepRelVertex and RelSubSet are the same as before, what do you mean by using the #computeDigest with the vertex id then ?
It's back to my comment about having the logic hardcoded into a single private static method which goes against OOP practices whereas instructing `HelpRexVertex` and `RelSubSet` instructing digest that the only field to consider for digest should be `id`

Reply:
Use the digest string to compute hashcode works well then the string is short, you can use the benchmark https://github.com/apache/calcite/pull/2034 to see the diff. The performance diff is negligible.